ui/src/components/tag/TagList.js

```diff
@@ -0,0 +1,483 @@
+            tokenExpiryMins: domain.tokenExpiryMins,
+            tagCertExpiryMins: domain.tagCertExpiryMins,
+            roleCertExpiryMins: domain.roleCertExpiryMins,
+            signAlgorithm: domain.signAlgorithm,
+            memberExpiryDays: domain.memberExpiryDays,
+            tagExpiryDays: domain.tagExpiryDays,
+            groupExpiryDays: domain.groupExpiryDays,
+            tags: domain.tags,
+        };
+    }
+
+    roleMetaObject(role) {
+        return {
+            selfServe: role.selfServe,
+            certExpiryMins: role.certExpiryMins,
+            reviewEnabled: role.reviewEnabled,
+            notifyRoles: role.notifyRoles,
+            serviceExpiryDays: role.serviceExpiryDays,
+            memberReviewDays: role.memberReviewDays,
+            serviceReviewDays: role.serviceReviewDays,
+            userAuthorityExpiration: role.userAuthorityExpiration,
+            userAuthorityFilter: role.userAuthorityFilter,
+            memberExpiryDays: role.memberExpiryDays,
+            tokenExpiryMins: role.tokenExpiryMins,
+            signAlgorithm: role.signAlgorithm,
+            groupExpiryDays: role.groupExpiryDays,
+            tags: role.tags,
+        };
+    }
+
+    updateMeta(meta, csrf, successMessage) {
+        this.api
+            .putMeta(
+                this.props.domain,
+                this.state.category === 'domain'
+                    ? this.props.domain
+                    : this.props.role,
+                meta,
+                'Updated ' + this.props.domain + ' Meta using Athenz UI',
+                csrf,
+                this.state.category
+            )
+            .then(() => {
+                this.reloadTags(successMessage, true);
+            })
+            .catch((err) => {
+                this.setState({
+                    errorMessage: RequestUtils.xhrErrorCheckHelper(err),
+                });
+            });
+    }
+
+    reloadTags(successMessage) {
+        if (this.state.category === 'domain') {
+            this.props.api
+                .getDomain(this.props.domain)
+                .then((data) => {
+                    this.updateStateAfterReload(data, successMessage);
+                })
+                .catch((err) => {
+                    this.setState({
+                        errorMessage: RequestUtils.xhrErrorCheckHelper(err),
+                    });
+                });
+        } else if (this.state.category === 'role') {
+            this.props.api
+                .getRole(this.props.domain, this.props.role)
+                .then((data) => {
+                    this.updateStateAfterReload(data, successMessage);
+                })
+                .catch((err) => {
+                    this.setState({
+                        errorMessage: RequestUtils.xhrErrorCheckHelper(err),
+                    });
+                });
+        }
+    }
+
+    updateStateAfterReload(data, successMessage) {
+        this.setState({
+            tags: data.tags,
```

Comment:
tags will be undefined after deleting last tag

Reply:
this is ok

Reply:
it throws an error, we should handle it similar to what we do in the constructor

Reply:
Handled that in `render` but it is better